### PR TITLE
Decode Julia profile

### PR DESCRIPTION
**Pull request by vchuravy**
Still WIP, and I am unsure if I got the finer details of the file-format right.
As an example a `CallFrame` is identifying a function right now and not a location inside a
function.

### CODE REVIEW

**Comment by davidanthoff** [comment]
Cool, thanks!

We can't use JSON3.jl, I'm afraid... We need something that works with JSON.jl v0.20.1, that is the version we'll have available in the VS Code extension.

The other thought I had in the initial implementation was to try to not allocate a new representation of the JSON output in memory, but instead just write/stream it out to JSON? But maybe that was silly... I had thought that might make sense because profile things could get very large? Not sure, we can probably ignore that.

**Comment by vchuravy** [comment]
Yeah I just didn't want to deal with writing JSON while having to deal with the more complicated parts ;)
Not sure yet if we can stream most of this data.

**Comment by davidanthoff** [comment]
I just tried it, and the values in `samples` look weird to me? When I export without C stuff, I get an array with only pointers to the node `1`?

**Comment by davidanthoff** [comment]
Actually, are you storing only references to root nodes in the `samples` vector? I had assumed it should be references to leaf nodes only?

**Comment by davidanthoff** [comment]
Looking at this:

image

I get the impression that the whole tree is up-side-down: it seems that as I drill down there, I'm getting the caller as the child node. I think it should be the other way around.

**Comment by davidanthoff** [comment]
I would also think that only leaf nodes should have tick info? Not sure, though...

I think you can take out all the `hitCount`, I don't believe the VS Code visualizer is using that.

**Comment by vchuravy** [comment]
> Actually, are you storing only references to root nodes in the samples vector? I had assumed it should be references to leaf nodes only?

I thought of it as the entry to the sample, but it being the exit makes more sense.

> I get the impression that the whole tree is up-side-down: it seems that as I drill down there, I'm getting the caller as the child node. I think it should be the other way around.

I have flipped the tree twice :P I think the first time around make more sense. It's weird since I am thinking about the profile data and what child means there, whereas I should probably think about what it means for the profile display.

**Comment by vchuravy** [comment]
I think the tree is in the right order. At least speedscope.app seems to think so. I fixed the samples bug. 

I get a spurious `randn` so there are probably still bugs and I can't open it on my copy of VSCode, due to the: " Cannot read property 'selfTime' of undefined."

**Comment by davidanthoff** [comment]
I can just open the file that gets generated with the latest Insiders edition, no problem... Weird...

The `samples` node still looks very weird to me. When I profile
```julia
function profile_test(n)
    for i = 1:n
        A = randn(100,100,20)
        m = maximum(A)
        Am = mapslices(sum, A; dims=2)
        B = A[:,:,5]
        Bsort = mapslices(sort, B; dims=1)
        b = rand(100)
        C = B.*b
    end
end
```
with `@profile profile_test(10)`, I only get five elements in the `samples` vector, that can't be right?

**Comment by vchuravy** [comment]
> with @profile profile_test(10), I only get five elements in the samples vector, that can't be right?

I got 57 samples. Are you on Windows? I think on windows the timer is limited to a magnitude lower sample rate.

Hm the VSCode extension gives me different results than the speedscope.app. Will comment on the main repo.

File: Project.toml [hunk 1]
```diff
@@ -8,7 +8,6 @@ JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
 Profile = "9abbd945-dff8-562f-b5e8-e1ebf5ef1b79"
 
 [compat]
-JSON = "0.20, 0.21"
 julia = "1"
```
Comment (vchuravy) [inline Project.toml:11]:
```suggestion
JSON = "0.20, 0.21"
julia = "1"
```

File: src/ChromeProfileFormat.jl [hunk 2]
```diff
@@ -2,123 +2,284 @@ module ChromeProfileFormat
...
+            frame.pointer == 0 && continue
+            frame.from_c && !from_c && continue
+
+            # A `frame` is an entry in a backtrace. A `ProfileNode` is a method in the call-graph
+            # which can be hit by multiple frames. So first we need to go from frame to method.
+            # As keys we will use `url` + `start_line`.
+            if frame.linfo === nothing # inlined frame
+                file = string(frame.file)
+                name = string(frame.func)
+                start_line = convert(Int64, frame.line) # TODO: Get start_line properly
```
Comment (vchuravy) [inline src/ChromeProfileFormat.jl:131]: @vtjnash This is where my question for getting the `start_line` of a method was motivated from.

**Comment by davidanthoff** [comment]
>  Are you on Windows?

Yes, that is probably it.

Should we merge this soon and then sort out any additional improvements in new PRs? Or do you want to continue iterating here? I'm indifferent :)

**Comment by vchuravy** [comment]
We can merge as is :)

File: src/ChromeProfileFormat.jl [hunk 3]
```diff
@@ -2,123 +2,284 @@ module ChromeProfileFormat
...
+            else
+                if frame.linfo isa Core.MethodInstance
+                    linfo = frame.linfo::Core.MethodInstance
+                elseif frame.linfo isa Core.CodeInfo
+                    # TODO: stackframes.jl describes this as a top-level frame
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/ChromeProfileFormat.jl:136]: It came from `eval`

File: src/ChromeProfileFormat.jl (hunk 2 above)
Comment (vtjnash) [reply src/ChromeProfileFormat.jl:131]: Means we couldn't compute it. Though if you want it, LLVM gave it to us, we just are choosing not to expose it (https://llvm.org/doxygen/classllvm_1_1DIContext.html)

**Comment by davidanthoff** [comment]
> Julia's Profile module now has separate computation of the tree (aka ProfileNode), so you might want to consider porting over to use that?

@vtjnash which versions of Julia would support that? We normally try to support all Julia 1.x versions in the VS Code extension... But maybe it would be ok to say that this new feature only works with newer versions of Julia...

**Pull request merged**